crates/ruff_python_parser/src/parser/statement.rs

```diff
@@ -2100,35 +2100,58 @@ impl<'src> Parser<'src> {
         }
     }
 
-    /// Parses any statement that is valid after an `async` token.
+    /// Parses a statement that is valid after an `async` token.
+    ///
+    /// If the statement is not a valid `async` statement, an error will be reported
+    /// and it will be parsed as a statement.
+    ///
     /// See:
-    ///  - <https://docs.python.org/3/reference/compound_stmts.html#grammar-token-python-grammar-async_with_stmt>
-    ///  - <https://docs.python.org/3/reference/compound_stmts.html#grammar-token-python-grammar-async_for_stmt>
-    ///  - <https://docs.python.org/3/reference/compound_stmts.html#grammar-token-python-grammar-async_funcdef>
+    /// - <https://docs.python.org/3/reference/compound_stmts.html#the-async-with-statement>
+    /// - <https://docs.python.org/3/reference/compound_stmts.html#the-async-for-statement>
+    /// - <https://docs.python.org/3/reference/compound_stmts.html#coroutine-function-definition>
     fn parse_async_statement(&mut self) -> Stmt {
         let async_start = self.node_start();
         self.bump(TokenKind::Async);
 
         match self.current_token_kind() {
+            // test_ok async_function_definition
+            // async def foo(): ...
             TokenKind::Def => Stmt::FunctionDef(ast::StmtFunctionDef {
                 is_async: true,
                 ..self.parse_function_definition(vec![], async_start)
             }),
+
+            // test_ok async_with_statement
+            // async with item: ...
             TokenKind::With => Stmt::With(ast::StmtWith {
                 is_async: true,
                 ..self.parse_with_statement(async_start)
             }),
+
+            // test_ok async_for_statement
+            // async for target in iter: ...
             TokenKind::For => Stmt::For(ast::StmtFor {
                 is_async: true,
                 ..self.parse_for_statement(async_start)
             }),
+
             kind => {
-                // Although this statement is not a valid `async` statement,
-                // we still parse it.
+                // test_err async_unexpected_token
+                // async class Foo: ...
+                // async while test: ...
+                // async x = 1
+                // async async def foo(): ...
+                // # TODO(dhruvmanila): Here, `match` is actually a Name token because
+                // # of the soft keyword # transformer
+                // async match test:
+                //     case _: ...
                 self.add_error(
```

Comment:
This recovery here slightly violates our recovery because we "drop" the async token. I wonder if we should instead parse this as a expression statement where `async` is an identifier.

```
async 
```

Reply:
It's probably seen in the existing tests as well. I think we shouldn't call `parse_statement` if we see a single `async` keyword. I think we should either:

* Do a lookahead and only parse `async` if we know we can parse out an async statement (a single `async` keyword than becomes an expression statement)
* Or, parse the `async` as an `ExpressionStatement` containing an `ASyncExpr` without any content.

The problem here is that we otherwise "drop" the `async` keyword. We bump the token when entering the function but then ignore it in the `parse_statement` call. You can see this for `async class X`. I think that should parse as a 

```
ExpressionStatement(AsyncExpression)
ClassDefinition
```

or 

```
ExpressionStatement(NameExpr(`async`))
ClassDefinition
```

but not as

```
ClassDefinition
```

because the `async` range then got completely omitted.